src/components/spatial/SpatialSubscriber.js

```diff
@@ -114,80 +116,119 @@ export default function SpatialSubscriber(props) {
 
   const [
     {
-      rasterLayersCallbacks,
+      imageLayerCallbacks,
+      segmentationLayerCallbacks,
     },
   ] = useAuxiliaryCoordination(
     COMPONENT_COORDINATION_TYPES.layerController,
     coordinationScopes,
   );
 
-  const use3d = rasterLayers?.some(l => l.use3d);
+  const use3d = imageLayers?.some(l => l.use3d);
 
-  const [urls, addUrl, resetUrls] = useUrls();
-  const [
-    isReady,
-    setItemIsReady,
-    setItemIsNotReady,
-    resetReadyItems,
-  ] = useReady(
-    SPATIAL_DATA_TYPES,
-  );
+  const [urls, addUrl] = useUrls(loaders, dataset);
   const [width, height, deckRef] = useDeckCanvasSize();
 
-  // Reset file URLs and loader progress when the dataset has changed.
-  // Also clear the array of automatically-initialized layers.
-  useEffect(() => {
-    resetUrls();
-    resetReadyItems();
-  // eslint-disable-next-line react-hooks/exhaustive-deps
-  }, [loaders, dataset]);
+  const [obsLabelsTypes, obsLabelsData] = useMultiObsLabels(
+    coordinationScopes, obsType, loaders, dataset, addUrl,
+  );
 
-  // Get data from loaders using the data hooks.
-  const [cells, cellsCount] = useCellsData(
-    loaders, dataset, setItemIsReady, addUrl, false,
-    { setSpatialSegmentationLayer: setCellsLayer },
-    { spatialSegmentationLayer: cellsLayer },
+  const hasExpressionData = useHasLoader(
+    loaders, dataset, DataType.OBS_FEATURE_MATRIX,
+    { obsType, featureType, featureValueType },
+    // TODO: get per-spatialLayerType expression data once #1240 is merged.
+  );
+  const hasCellsData = useHasLoader(
+    loaders, dataset, DataType.OBS_SEGMENTATIONS,
+    { obsType: 'cell' }, // TODO: use obsType in matchOn once #1240 is merged.
```

Comment:
Are you waiting on me to review that? I was unclear where we left that.  I can have a look at it again if it's blocking.